internal/torcx/remote.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2018 CoreOS Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package torcx
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/euank/gotmpl"
+	"github.com/pkg/errors"
+)
+
+var (
+	errNilRemote          = errors.New("nil Remote")
+	errEmptyUsrMountpoint = errors.New("empty USR mountpoint")
+	errEmptyTemplateURL   = errors.New("empty remote URL template")
+)
+
+// evaluateURL evaluates the URL template for a remote
+// and performs variables substitution sourcing values from
+// `/etc/os-release`.
+func (r *Remote) evaluateURL(usrMountpoint string) (string, error) {
+	if r == nil {
+		return "", errNilRemote
+	}
+	if usrMountpoint == "" {
+		return "", errEmptyUsrMountpoint
+	}
+	if r.TemplateURL == "" {
+		return "", errEmptyTemplateURL
+	}
+
+	if !needSubstitution(r.TemplateURL) {
+		return r.TemplateURL, nil
+	}
+
+	osReleasePath := VendorOsReleasePath(usrMountpoint)
+	fp, err := os.Open(osReleasePath)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to open %s", osReleasePath)
+	}
+	defer fp.Close()
+	osMeta, err := parseOsRelease(fp)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to parse %s", osReleasePath)
+	}
+	osMeta["COREOS_USR"] = usrMountpoint
```

Comment:
I don't understand the value of exposing the `usrMountpoint` to the template logic.

I understand it's needed by this function to find the appropriate version (e.g. if `/usr` is mounted by `update_engine` at `/tmp/au_postint_mount.xxx`, it obviously needs to read os-release from there), but I don't see how the remote url would ever need to care about this detail.

Did you have any specific cases or use in mind? If there's not a specific use-case compelling this, I'd rather not support it.

Reply:
This is something that I missed in the original design doc, so here is the full rationale. The current and upcoming USR partition both carries a local remote (`com.coreos.cl`). The manifest for that will contain a URL like `file:///usr/share/torcx/remotes/...`. Such URL works for the current USR, but at update-engine we also need to inspect the remote coming with next-OS, which is mounted under a temporary dir. Options we have are:
 * introduce a template variable, and point the URL to `file://${COREOS_USR}/share/torcx/...`
 * manually replace `/usr` when processing the URL from `com.coreos.cl`

Both are a bit of corner cases, but I felt that the first one would be a bit less surprising.

Reply:
Thanks for explaining, I think it makes sense to have that template variable now.

However, I think the `base_url` bit needs to also include `file` as a supported protocol, and ideally should have an example explaining this possibility (if only to say 'this is supported, but generally should only be used by pseudo-remotes distributed by the creator/vendor of /usr').

I think having that clarification could have also resolved my confusion.

Reply:
Indeed. I'll fix and add a couple of explicit notes to `remote-manifest-v0.md`.